### DISCUSSION TITLE
How to preserve generic type in the constraint of a VarType?

### DISCUSSION THREAD

**Discussion by diego-pm** [microsoft/pyright#4369] [Q&A]
Take the following example, where I define a TypeVar `_R` constrained to a list or set:

```python
_R = TypeVar("_R", set, list)

def foo(x: _R) -> _R:
    return x

x = [1, 2, 3]  # type = list[int]
res = foo(x)  # type = list[Any]
```

I would like that the function `foo` preserves the generic type of the input. In the example, pyright infers that the type of `res` is `list[Any]`. How can I make that the elements have the same type (`list[int]`) ?

I have tryied to include the generic variable to the list and set:

```python
_T = TypeVar("_T")
_R = TypeVar("_R", set[_T], list[_T])
```
But I get the error: TypeVar constraint type cannot be generic.

**Comment by erictraut** [comment]
A constrained TypeVar is probably not the right tool for the job you're trying to do. An overload is probably a better approach.

```python
_T = TypeVar("_T")

@overload
def foo(x: set[_T]) -> set[_T]: ...
@overload
def foo(x: list[_T]) -> list[_T]: ...
def foo(x: set[_T] | list[_T]) -> set[_T] | list[_T]:
    return x
```

**Reply by diego-pm** [reply]
Thanks for the response! What if I have this problem with a generic class? I can not use overload directly:

```python
class Bar(Generic[_R]):
    def __init__(x: _R):
        self.x = x
    def __iter__() -> Generator[_R, None, None]:
        yield self.x
```

I would the similar application, where `next(iter(Bar([1, 2, 3])))` has type `list[int]`.

**Reply by erictraut** [reply]
Do you require that your generic class preserve the fact that that the argument passed to the constructor is a `set` or a `list`? Or does your class support any `Iterable`? The latter is much more common. 

```python
from typing import Generator, Generic, TypeVar, Iterable

_T = TypeVar("_T")

class Bar(Generic[_T]):
    def __init__(self, x: Iterable[_T]):
        self.x = x

    def __iter__(self) -> Generator[Iterable[_T], None, None]:
        yield self.x
```

I don't see a good way to preserve the `list[T]` or `set[T]` using a constrained TypeVar scoped to a class. You're attempting to push the type system beyond what it was designed to do. There is some discussion in the typing community of supporting generic bounds for TypeVars — and more generally the concept of a "higher-kinded type" (HKT) which allows TypeVars that are generic on other TypeVars, but the effort to define this feature are still very nascent. As far as I know, no one has explored what it would mean to allow for generic constraint types in a constrained TypeVar. The semantics for a constrained TypeVar are already very strange. They don't exist in any other type system that I'm aware of; they're an oddity of Python.

**Reply by diego-pm** [reply]
I can not just use an iterable in my use case, `list` and `set` are really `list` or `pd.Series` or `pd.DataFrame`. The code if more like this:

```python
_R = TypeVar("_R", list, pd.Series, pd.DataFrame)

class Bar(Generic[_R]):
    def __init__(x: _R):
        self.x = x
    def __iter__() -> Generator[_R, None, None]:
        for i in range(0, len(self.x), 10):
            if isinstance(self.x, (pd.Series, pd.DataFrame)):
                yield self.x.iloc[i:i+10]
            else:
                yield self.x[i:i+10]
```
